rusthotp/src/lib.rs

```diff
@@ -2,35 +2,54 @@ extern crate hmacsha1;
-/// stages, as per [RFC 4226](https://www.ietf.org/rfc/rfc4226.txt):
+/// This implementation of HMAC-Based One Time Passwords proceeds in three stages, as per [RFC
+/// 4226](https://www.ietf.org/rfc/rfc4226.txt):
 ///
-/// 1. Generate an HMAC-SHA-1 value given a shared secret `key` and a
-/// moving factor `counter` that *should* change on every use
-/// 2. Extraction of a 4-byte value from the SHA-1 digest
-/// 3. Deriving a final HOTP value between 0 and
-/// 10^{`desired_code_length`}-1 from the 4-byte value extracted
-/// in the last step.
+/// 1. Generate an HMAC-SHA-1 value given a shared secret `key` and a moving factor `counter` that *should* change on
+///    every use
+/// 2. Extraction of an indexing offset by reading the 4 low-order bits of the last byte of the SHA-1 digest
+/// 3. Extraction of a 4-byte value from the SHA-1 digest, starting at the index offset determined by the previous step
+/// 4. Deriving a final HOTP value between 0 and 10^{`desired_code_length`} -1 from the 4-byte value extracted in the
+///    previous step.
 ///
-/// Note that it is the responsibility of the caller to increment the counter accordingly per invocation.
+/// Note that it is the responsibility of the caller to increment the counter accordingly per invocation. Consider the
+/// use of the [byteorder crate](https://crates.io/crates/byteorder) if you need to convert from a numeric type into the
+/// array of 8 bytes that are expected by `hotp()`.
 ///
 /// # Examples
 ///
-/// To generate a six-digit HOTP value between 0 and 10^6 - 1, given the shared
-/// secret ASCII key `"Hello world!"` and current counter `0`:
+/// To generate a six-digit HOTP value (i.e. between 0 and 10^6 - 1), given the shared secret ASCII key `"Hello world!"`
+/// and current counter `0`:
 ///
 /// ```
 /// let six_digit_result = rusthotp::hotp(6, "Hello world!".as_bytes(), &[0; 8]);
 /// assert_eq!(format!("{}", six_digit_result), "124111");
 /// ```
 ///
-/// To generate an eight-digit HOTP value between 0 and 10^8 - 1, given the shared
-/// secret ASCII key `"12345678901234567890"` and current counter `1`:
+/// To generate an eight-digit HOTP value (i.e. between 0 and 10^8 - 1), given the shared secret ASCII key
+/// `"12345678901234567890"` and current counter `1`:
 ///
 /// ```
 /// let eight_digit_result = rusthotp::hotp(8, "12345678901234567890".as_bytes(), &[0, 0, 0, 0, 0, 0, 0, 1]);
 /// assert_eq!(format!("{}", eight_digit_result), "94287082");
 /// ```
+///
+/// To generate a six-digit HOTP value given the shared secret ASCII key "Hello world!" and the unix timestamp
+/// 1493006116 as the counter:
+///
+/// ```
+/// # extern crate byteorder;
+/// # extern crate rusthotp;
+/// # fn main() {
+/// use byteorder::{BigEndian, ByteOrder};
+/// let timestamp : i64 = 1493006116;
+/// let mut unpacked_timestamp = [0; 8];
+/// BigEndian::write_i64(&mut unpacked_timestamp, timestamp);
+/// let result = rusthotp::hotp(6, "Hello world!".as_bytes(), &unpacked_timestamp);
```

Comment:
My intent was to keep the hotp library as generic as possible. That is, RFC 4226 defines `C` as an "8-byte counter value", and I think it makes sense at this level of abstraction to be as true as possible to the specification.

I think it's kind of important to view the counter input to `hotp()` as a arbitrary string of binary data rather than a meaningful type; a u64 implies that the input data is an integer, but from the perspective of the algorithm, it isn't. We aren't performing any numeric operations on it, we're just passing a blob of data into a hashing function.

Finally, altering `hotp()` to take a u64 doesn't even help us implement RFC6238, the most common client. TOTP uses a unix timestamp, which is a _signed_ 64 bit integer. That timestamp is divided by the timestep and then passed into `hotp()`, at which point we have to deliberately discard the signedness of the type and reinterpret it as a binary blob anyway.